coderd/provisionerdserver/provisionerdserver_test.go

```diff
@@ -94,29 +97,53 @@ func TestAcquireJobWithCancel_Cancel(t *testing.T) {
 func TestHeartbeat(t *testing.T) {
 	t.Parallel()
 
-	numBeats := 3
-	ctx := testutil.Context(t, testutil.WaitShort)
-	heartbeatChan := make(chan struct{})
-	heartbeatFn := func(hbCtx context.Context) error {
-		t.Logf("heartbeat")
-		select {
-		case <-hbCtx.Done():
-			return hbCtx.Err()
-		default:
-			heartbeatChan <- struct{}{}
-			return nil
-		}
-	}
+	var (
+		ctx, cancel    = context.WithTimeout(context.Background(), testutil.WaitShort)
+		mClock         = clock.NewMock(t)
+		trapHbLoop     = mClock.Trap().NewTicker("heartbeat_loop")
+		trapHbLastSeen = mClock.Trap().Now("update_provisioner_daemon_last_seen_at")
+		ctrl           = gomock.NewController(t)
+		mStore         = dbmock.NewMockStore(ctrl)
+		orgID          = uuid.New()
+		heartbeatDone  = make(chan struct{})
+	)
+	defer cancel()
+	mStore.EXPECT().GetDefaultOrganization(gomock.Any()).AnyTimes().Return(database.Organization{ID: orgID}, nil)
+	mStore.EXPECT().UpsertProvisionerDaemon(gomock.Any(), gomock.Any()).AnyTimes().Return(database.ProvisionerDaemon{
+		Tags: map[string]string{provisionersdk.TagScope: provisionersdk.ScopeOrganization},
+	}, nil)
 	//nolint:dogsled
 	_, _, _, _ = setup(t, false, &overrides{
 		ctx:               ctx,
-		heartbeatFn:       heartbeatFn,
+		clock:             mClock,
+		store:             mStore,
 		heartbeatInterval: testutil.IntervalFast,
+		heartbeatDone:     heartbeatDone,
 	})
 
-	for i := 0; i < numBeats; i++ {
-		testutil.RequireRecvCtx(ctx, t, heartbeatChan)
+	// Wait for heartbeat loop to start
+	trapHbLoop.MustWait(ctx).Release()
+
+	// Simulate some heartbeats
+	for i := 0; i < 10; i++ {
+		// Advance to the next scheduled interval
+		_, _ = mClock.AdvanceNext()
+		// Wait for the call to Now() in heartbeat default select case
+		call := trapHbLastSeen.MustWait(ctx)
+		// EXPECT the call
+		mStore.EXPECT().UpdateProvisionerDaemonLastSeenAt(gomock.Any(), gomock.Any()).Times(1).Return(nil)
+		// Start the clock again and wait for things to happen!
+		call.Release()
 	}
+
+	cancel()
+	// heartbeat may or may not be waiting to update last seen at
+	mStore.EXPECT().UpdateProvisionerDaemonLastSeenAt(gomock.Any(), gomock.Any()).MinTimes(0).Return(context.Canceled)
+	// Close the traps and let time continue
+	trapHbLoop.Close()
+	trapHbLastSeen.Close()
+	// We should be done now
+	<-heartbeatDone
```

Comment:
I don't think we need this `heartbeatDone` thing at all.  `goleak` for checking for leaked go routines, and if you use `clock.TickerFunc()` then by design it cleans up when the context is canceled. No need to test that here.